### DISCUSSION TITLE
Error Building JoltPhysics as Shared Library on MinGW/GCC (undefined reference to __emutls_v)

### DISCUSSION THREAD

**Discussion by Zh0ny** [jrouwe/JoltPhysics#1266] [Q&A] [answered]
Hello, I'm trying to build [JoltPhysicsHelloWorld](https://github.com/jrouwe/JoltPhysicsHelloWorld) with Jolt Physics as a shared library, but I'm encountering the following error:
```
[ 99%] Linking CXX shared library libJolt.dll
[ 99%] Built target Jolt
[ 99%] Building CXX object CMakeFiles/HelloWorld.dir/D_/Dev/cpp/JoltPhysicsHelloWorld/Source/HelloWorld.cpp.obj
[100%] Linking CXX executable HelloWorld.exe
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/14.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\HelloWorld.dir/objects.a(HelloWorld.cpp.obj):HelloWorld.cpp:(.rdata$.refptr.__emutls_v._ZN3JPH11PhysicsLock6sLocksE[.refptr.__emutls_v._ZN3JPH11PhysicsLock6sLocksE]+0x0): undefined reference to `__emutls_v._ZN3JPH11PhysicsLock6sLocksE'
collect2.exe: error: ld returned 1 exit status
mingw32-make[2]: *** [CMakeFiles\HelloWorld.dir\build.make:100: HelloWorld.exe] Error 1
mingw32-make[1]: *** [CMakeFiles\Makefile2:99: CMakeFiles/HelloWorld.dir/all] Error 2
mingw32-make: *** [Makefile:135: all] Error 2
```
You can see the full build output in the attached log: [build_output.log](https://github.com/user-attachments/files/17051220/build_output.log)

compiler: [winlibs-x86_64-posix-seh-gcc-14.2.0-llvm-18.1.8-mingw-w64ucrt-12.0.0-r1](https://winlibs.com/)

Steps I followed:

1º Cloned the repository: https://github.com/jrouwe/JoltPhysicsHelloWorld.git
2º Added cmake_windows_mingw.sh from JoltPhysics to the Build folder and added the following instructions:
```
export CC="C:/mingw64/bin/gcc.exe"
export CXX="C:/mingw64/bin/g++.exe"
```
3º Opened Git Bash in the folder and ran: ./cmake_windows_mingw.sh
4º Opened /MinGW_Debug/_deps/joltphysics-src/Build and set BUILD_SHARED_LIBS to ON in CMakeLists.txt.
5º In Git Bash, executed the build command
6º After building, I encountered this error at the end:
```
[100%] Linking CXX executable HelloWorld.exe
C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/14.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\HelloWorld.dir/objects.a(HelloWorld.cpp.obj):HelloWorld.cpp:(.rdata$.refptr.__emutls_v._ZN3JPH11PhysicsLock6sLocksE[.refptr.__emutls_v._ZN3JPH11PhysicsLock6sLocksE]+0x0): undefined reference to `__emutls_v._ZN3JPH11PhysicsLock6sLocksE'
collect2.exe: error: ld returned 1 exit status
mingw32-make[2]: *** [CMakeFiles\HelloWorld.dir\build.make:100: HelloWorld.exe] Error 1
mingw32-make[1]: *** [CMakeFiles\Makefile2:99: CMakeFiles/HelloWorld.dir/all] Error 2
mingw32-make: *** [Makefile:135: all] Error 2
```

Alternative Steps:

1º Cloned the repository: https://github.com/jrouwe/JoltPhysicsHelloWorld.git
2º Opened the folder in VSCode with the CMake Tools extension installed
3º Configured the CMakeLists.txt in the Build folder using CMake Tools.
4º Opened Build\out\build\GCC 14.2.0 x86_64-w64-mingw32\_deps\joltphysics-src\Build\ and set BUILD_SHARED_LIBS to ON in CMakeLists.txt 
5º From Build\out\build\GCC 14.2.0 x86_64-w64-mingw32, excluded CMakeCache.txt and configured again.
6º Built the project using CMake Tools' build option, but I got the same error:
```
[build] [ 99%] Linking CXX shared library libJolt.dll
[build] [ 99%] Built target Jolt
[build] [ 99%] Building CXX object CMakeFiles/HelloWorld.dir/D_/Dev/cpp/JoltPhysicsHelloWorld/Source/HelloWorld.cpp.obj
[build] [100%] Linking CXX executable HelloWorld.exe
[build] C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/14.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\HelloWorld.dir/objects.a(HelloWorld.cpp.obj):HelloWorld.cpp:(.rdata$.refptr.__emutls_v._ZN3JPH11PhysicsLock6sLocksE[.refptr.__emutls_v._ZN3JPH11PhysicsLock6sLocksE]+0x0): undefined reference to `__emutls_v._ZN3JPH11PhysicsLock6sLocksE'
[build] collect2.exe: error: ld returned 1 exit status
[build] mingw32-make[3]: *** [CMakeFiles\HelloWorld.dir\build.make:100: HelloWorld.exe] Error 1
[build] mingw32-make[2]: *** [CMakeFiles\Makefile2:99: CMakeFiles/HelloWorld.dir/all] Error 2
[build] mingw32-make[1]: *** [CMakeFiles\Makefile2:106: CMakeFiles/HelloWorld.dir/rule] Error 2
[build] mingw32-make: *** [Makefile:168: HelloWorld] Error 2
```

Can anyone help me resolve this issue?

**Comment by jrouwe** [comment]
I've created a fix for this in #1273. I'm seeing more issues though with building as a shared lib with MingW (the unit tests also don't link), so this may not be the last error you're going to see.

**Comment by jrouwe** [comment, marked as answer]
With #1274 the issue should be fully fixed now.

**Reply by Zh0ny** [reply]
;-; Thank you so much for your attention with this issue! I'm greatly appreciate the time and effort you’ve put into helping the community. Your support has been invaluable, I'm already try again with the changes and everything is working well.
Sorry for not replying to your first response right away. When you mentioned some other errors, I started looking into how to fix them. 

Thanks again!
